### ISSUE TITLE
Use cookies only if allowed

### ISSUE THREAD

**Issue by lekoala** [tractorcow-farm/silverstripe-fluent#678]
While working on my module https://github.com/lekoala/silverstripe-cookieconsent I considered the case of the FluentLocale cookie. Most websites now require cookies to be accepted before setting any cookie. Since the cookie is set by a middleware, it is difficult to update the config flag based on user preference. For instance, I tried something like this

       $class = \TractorCow\Fluent\Middleware\DetectLocaleMiddleware::class;
        if (!class_exists($class)) {
            return;
        }
        $v = $class::config()->persist_cookie;

        // If we choose to persist cookies, we should check our cookie consent first
        if ($v) {
            $status = Cookie::get('cookieconsent_status');
            if ($status && $status != 'allow') {
                $dont_persist = true;
            }

            $status = Cookie::get('cookie_consent_user_accepted');
            if ($status != 'true') {
                $dont_persist = true;
            }
        }
        if ($dont_persist) {
                Config::modify()->set($class, 'persist_cookie', false);
            }

But it doesn't work because it gets called after the middlewares. This means I would need to add a middleware before fluent just to prevent that, and that feels clunky.

Would it be possible possible to either:
- execute this later so that earlier code can modify the persist_cookie value if needed (set it to false if cookies are not accepted yet)
- disable the middleware altogether and recommend developers to call a method to persist cookies if they are allowed?

What do you think? I'm going for now to basically set persist_cookie to false and manually persist them in my controllers I guess that's not how it should be done :-)

**Comment by tractorcow** [comment]
It looks like "strictly necessary" (I.e. cookies that the site require to function) do not require consent.

https://www.privacypolicies.com/blog/eu-cookie-law/

I think in this case, it still adds value to make sure that the cookie usage of any site is disclosed to any site user, but it's not necessary to block it based on user consent.

If that were the case, you'd need to apply the same logic to the PHP session cookie, so only applying this to fluent doesn't really seem consistent.

I would be happy to add a disclaimer / note to the fluent readme to provide peace of mind to developers using this module.

**Comment by lekoala** [comment]
From what I've read, but I may be wrong, it's not strictly necessary since you can avoid it and rely on the url only.
Therefore, it's only a matter of remembering your preferences. If that's the case, it's seen as a functional cookies. If i take the examples from cookieconsent.com, there are four levels of cookies:
- strictly necessary: session cookies for example
- functional : These cookies are used to provide you with a more personalized experience on our website and to remember choices you make when you use our website. For example, we may use functionality cookies to remember your language preferences or remember your login details. <-- i think that's the category for fluent
- tracking
- advertising

It's only strictly necessary, for example, in the cms, where you don't have the url to help you. Even then, you have the session, so you can get away without a cookie as well.

Regardless of law interpretations, cookies scanner like https://www.cookiebot.com/fr/ will pick up fluent cookie by default and then you need endless talks with your client to make him understand that this whole cookie thing is not very reasonable in the very place :-) so really, what I do now is this : https://github.com/lekoala/silverstripe-multilingual/blob/master/src/LangHelper.php#L77 it works well and it doesn't have many draw back. if someone really doesn't want to remember his language preference, so be it :-)

What do you think?

**Comment by tractorcow** [comment]
If we can rely on the php session, I guess we could safely assume that would handle the main frontend case. in which case, may as well document how to turn it off for a site? (rather than conditionally turning it on at all).

**Comment by lekoala** [comment]
indeed

so my proposal, for now, which would require minimal code change would be to check another cookie value before persisting the id. It would be opt in (meaning : no change if it's not configured). It means adding two new config keys : consent_cookie_name, consent_cookie_value. If consent_cookie_name is set, it will check it's value (could default to true/allow if none set) before trying to persist the cookie.

would be happy to make a PR with that.

**Comment by RoyalPulp** [comment]
Hi @lekoala and @tractorcow 
I'd like to know, if there is any progress here? I just had a conversation with a privacy specialist about cookies and german/european privacy law - and the fluent cookies seem to be an issue for a german based website.
Thanks for any progress!

**Comment by lekoala** [comment]
@RoyalPulp nope, no progress but you can disable persist_cookie and do it only if cookies are accepted

**Comment by tractorcow** [comment]
If all of your routes are unique you should be able to rely on routing to determine the locale instead of cookies. By utilising the multi-domain feature of fluent you can have multiple top level home pages for each locale as well, one for each domain, and never have to rely on cookies at all.
